dask/blockwise.py

```diff
@@ -436,10 +437,25 @@ def optimize_blockwise(graph, keys=()):
     --------
     rewrite_blockwise
     """
-    out = _optimize_blockwise(graph, keys=keys)
-    while out.dependencies != graph.dependencies:
-        graph = out
+    with warnings.catch_warnings():
```

Comment:
A sanity check here would be appreciated.

Something like this causes the warning

```python
In [2]: x = da.arange(10, chunks=5)

In [3]: y = np.arange(10)

In [4]: x.map_blocks(lambda x, y: x, 'abc').map_blocks(lambda x, y: x, y).compute()
/Users/taugspurger/sandbox/dask/dask/blockwise.py:590: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  contains = index in indices
Out[4]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```

The relevant section

```pytb
In [8]: debug
> /Users/taugspurger/sandbox/dask/dask/blockwise.py(590)rewrite_blockwise()
    588             for i, index in enumerate(new_indices):
    589                 try:
--> 590                     contains = index in indices
    591                 except (ValueError, TypeError):
    592                     contains = False

ipdb> index
('arange-809ffb1d9ea3bf099ca5d2fe65db1290', ('.0',))
ipdb> indices
[(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]), None)]
```

I think I convinced myself that this will continue to return False. IIUC, the error from comparing the scalar string 'arange-...' to the ndarray

```python
In [5]: 'abc' == np.arange(4)
/Users/taugspurger/.virtualenvs/dask-dev/bin/ipython:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  #!/Users/taugspurger/Envs/dask-dev/bin/python3
Out[5]: False
```

Now that returns false, In the future it would be an array of False. Past this, I'm less sure of what will happen, but I think we'll be OK, since (I think) the `in` comparison will still evaluate to False.

Reply:
> Now that returns false, In the future it would be an array of False. Past this, I'm less sure of what will happen, but I think we'll be OK, since (I think) the in comparison will still evaluate to False.

Will it?  Or will the array of Falses be truthy and so return True?  (I don't know which will happen).

Reply:
So for an operation like

```python
index = ('a', 1)
indicies = [('b', 1)]
index in indices
```

I think the order is something like

```
any(index == item for item in indicies) -> any([('a', 1) == ('b', 1)])
```

which uses tuple equality, which I think does `all(element wise equality)`

```
any([
    all('a'== x and 1 == y)
    for x, y in indices
]) -> any(all([False, True]) -> False
``` 

Now, if we change indices to `[(np.array([1, 2]), 1)]`, like in the real problem, that last expression becomes, in the future,

```
any(all([np.array([False, False]), True]))
```

Which would raise a ValueError (again, in the future).

---

In summary I was wrong about the comparison still returning False, but right that we'll still get the same answer, since we catch ValueError and TypeError. I'll update the docstring.